### PR TITLE
fix: remove price history data that is out of bounds of a given timeframe

### LINKED ISSUE #8093: Weird charts showing intermittently for assets on asset page
### Overview

When viewing assets on the asset page, getting some weird graphs. Not sure how it gets into the state. Clear cache seems to work but then the graphs can play up again. 

### References and additional details

image
image
image

### Acceptance Criteria

graph should be accurate representation of price

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Ownership

- [X] If my bounty needs engineering or needs product I have added the respective labels on the right
- [X] As the sponsor of this bounty I will review the changes in a preview environment (ops/product) or review the PR (engineering)

### Estimated effort

_No response_

### Sponsor / Stakeholder

_No response_

### Bounty Hunters

- [Join our discord](https://discord.gg/shapeshift)
- Include an expected timeline for you to complete work in the work plan when you apply for this bounty!
- Please refer to [this link](https://shapeshiftdao.gitbook.io/getting-started/how-to-contribute/bounties#resources-for-new-contributors-and-bounty-hunters) for some basic info
- Please ***do not*** start work on this issue until you are approved in Gitcoin.

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Fixes corrupted price charts caused by price samples spanning times outside of the selected timeframe. For example for the 24 hour time period, data prior to 24 hours ago was being displayed, resulting in corrupted charts in the UI. The solution was to trim price history data when it's upserted in redux, and also when it's read from redux.

NOTE: It's possible there is still chart corruption occurring due to incorrectly sorted price history samples. If this is the case, let me know and I'll add sorting to the selector. However if this is not the case we'd be best off avoiding the additional compute.

## Issue (if applicable)

closes #8093

## Risk

> High Risk PRs Require 2 approvals

Low risk, this affects display of price charts only, which is already broken.

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

## Screenshots (if applicable)

Before:

After:

### CODE REVIEW

File: src/state/slices/marketDataSlice/utils.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import type { AssetId } from '@shapeshiftoss/caip'
+import type { HistoryData, HistoryTimeframe, PartialRecord } from '@shapeshiftoss/types'
+import { getHistoryTimeframeBounds } from '@shapeshiftoss/utils'
+import { merge } from 'lodash'
+import type { SupportedFiatCurrencies } from 'lib/market-service'
+
+import type { PriceHistoryByTimeframe } from './types'
+
+export const getTrimmedOutOfBoundsMarketData = <T extends SupportedFiatCurrencies | AssetId>(
```
Comment (woodenfurniture) [inline src/state/slices/marketDataSlice/utils.ts:9]: Creates a new array - used in selectors

File: src/state/slices/marketDataSlice/utils.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    if (!idHistory) continue
+    results[id] = idHistory.filter(
+      ({ date }) => date >= startTimeStampMillis && date <= endTimeStampMillis,
+    )
+  }
+
+  return results
+}
+
+export const trimOutOfBoundsMarketData = <T extends SupportedFiatCurrencies | AssetId>(
```
Comment (woodenfurniture) [inline src/state/slices/marketDataSlice/utils.ts:35]: Mutates the input - used in actions

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Confirmed this fixes the individual assets I have with corrupted chart state 👌 

Noticed one potential issues with the overall balance chart:

Screenshot 2024-12-02 at 10 14 57

Today, for e.g. I seem to have no history before 0000 local time. Feels a little like a bug related to "current day" (in this PR) vs "last 24 hours" (expected).

**Comment by woodenfurniture** [comment]
> Today, for e.g. I seem to have no history before 0000 local time. Feels a little like a bug related to "current day" (in this PR) vs "last 24 hours" (expected).

@0xApotheosis right you are, fixed!




And comparing the chart against coingeck:



6bc641a31f87738fbb6e286f5c062d43d4081de7

**Review by 0xApotheosis — APPROVED** [review]
Total balance bug fixed 👌 

Note, I'm still able to get broken charts if I don't clear my cache - I'm not sure if that's expected. That said, it's significantly better than the current `develop` state.

This branch:

Screenshot 2024-12-02 at 12 41 47

`develop`:

Screenshot 2024-12-02 at 12 56 32

File: packages/utils/src/historyTimeframe.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import { HistoryTimeframe } from '@shapeshiftoss/types'
+import dayjs from 'dayjs'
+
+import { assertUnreachable } from './assertUnreachable'
+
+export const getHistoryTimeframeBounds = (
+  timeframe: HistoryTimeframe,
+  end = dayjs().endOf('day'),
+) => {
+  let start = end // Shut up typescript.
```
Comment (gomesalexandre) [inline packages/utils/src/historyTimeframe.ts:10]: 😄

**Comment by woodenfurniture** [comment]
> Note, I'm still able to get broken charts if I don't clear my cache - I'm not sure if that's expected. That said, it's significantly better than the current develop state.

I wasn't able to repro likely due to different local state, though here's a fix bab796ff4902dfc7c419379704686e33507e6efb (hopefully)

**Comment by 0xApotheosis** [comment]
Confirming https://github.com/shapeshift/web/pull/8235/commits/bab796ff4902dfc7c419379704686e33507e6efb fixed it 👌 

Screenshot 2024-12-02 at 13 59 50

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Tested locally against develop (left), LGTM minus one q:

- Assets price chart is fixed ❔ 

https://jam.dev/c/c0b7b30d-9adc-4274-9063-26e740716444

@woodenfurniture noting that "All" price chart for e.g BTC only goes up to 2 years ago, vs. all the way to 2013 in develop, this looks like a regression, is this known about / expected? 

This can be repro'd for other dinosaur coins too e.g CharlieCoin:

image
image

Balance chart looks sane:

image
image
image
image

Noting:

- 24h chart looks v. different (perhaps because of degraded accounts?) as well as 1M one showing green vs. red previously 
- 1h and all are perma loading
image
image

**Pull request merged**